File: src/services/skills.ts
import { invokeTauriOrNull } from "./tauriInvoke";
import type { CliKey } from "./providers";

export type SkillRepoSummary = {
  id: number;
  git_url: string;
  branch: string;
  enabled: boolean;
  created_at: number;
  updated_at: number;
};

export type InstalledSkillSummary = {
  id: number;
  skill_key: string;
  name: string;
  description: string;
  source_git_url: string;
  source_branch: string;
  source_subdir: string;
  enabled_claude: boolean;
  enabled_codex: boolean;
  enabled_gemini: boolean;
  created_at: number;
  updated_at: number;
};

export type AvailableSkillSummary = {
  name: string;
  description: string;
  source_git_url: string;
  source_branch: string;
  source_subdir: string;
  installed: boolean;
};

export type SkillsPaths = {
  ssot_dir: string;
  repos_dir: string;
  cli_dir: string;
};

export type LocalSkillSummary = {
  dir_name: string;
  path: string;
  name: string;
  description: string;
};

export async function skillReposList() {
  return invokeTauriOrNull<SkillRepoSummary[]>("skill_repos_list");
}

export async function skillRepoUpsert(input: {
  repo_id?: number | null;
  git_url: string;
  branch: string;
  enabled: boolean;
}) {
  return invokeTauriOrNull<SkillRepoSummary>("skill_repo_upsert", {
    repoId: input.repo_id ?? null,
    gitUrl: input.git_url,
    branch: input.branch,
    enabled: input.enabled,
  });
}

export async function skillRepoDelete(repoId: number) {
  return invokeTauriOrNull<boolean>("skill_repo_delete", { repoId });
}

export async function skillsInstalledList() {
  return invokeTauriOrNull<InstalledSkillSummary[]>("skills_installed_list");
}

export async function skillsDiscoverAvailable(refresh: boolean) {
  return invokeTauriOrNull<AvailableSkillSummary[]>("skills_discover_available", {
    refresh,
  });
}

export async function skillInstall(input: {
  git_url: string;
  branch: string;
  source_subdir: string;
  enabled_claude: boolean;
  enabled_codex: boolean;
  enabled_gemini: boolean;
}) {
  return invokeTauriOrNull<InstalledSkillSummary>("skill_install", {
    gitUrl: input.git_url,
    branch: input.branch,
    sourceSubdir: input.source_subdir,
    enabledClaude: input.enabled_claude,
    enabledCodex: input.enabled_codex,
    enabledGemini: input.enabled_gemini,
  });
}

export async function skillSetEnabled(input: {
  skill_id: number;
  cli_key: CliKey;
  enabled: boolean;
}) {
  return invokeTauriOrNull<InstalledSkillSummary>("skill_set_enabled", {
    skillId: input.skill_id,
    cliKey: input.cli_key,
    enabled: input.enabled,
  });
}

export async function skillUninstall(skillId: number) {
  return invokeTauriOrNull<boolean>("skill_uninstall", { skillId });
}

export async function skillsLocalList(cliKey: CliKey) {
  return invokeTauriOrNull<LocalSkillSummary[]>("skills_local_list", { cliKey });
}

export async function skillImportLocal(input: { cli_key: CliKey; dir_name: string }) {
  return invokeTauriOrNull<InstalledSkillSummary>("skill_import_local", {
    cliKey: input.cli_key,
    dirName: input.dir_name,
  });
}

export async function skillsPathsGet(cliKey: CliKey) {
  return invokeTauriOrNull<SkillsPaths>("skills_paths_get", { cliKey });
}

File: src-tauri/src/commands/skills.rs
//! Usage: Skills management related Tauri commands.

use crate::app_state::{ensure_db_ready, DbInitState};
use crate::{blocking, skills};

#[tauri::command]
pub(crate) async fn skill_repos_list(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
) -> Result<Vec<skills::SkillRepoSummary>, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("skill_repos_list", move || skills::repos_list(&app)).await
}

#[tauri::command]
pub(crate) async fn skill_repo_upsert(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    repo_id: Option<i64>,
    git_url: String,
    branch: String,
    enabled: bool,
) -> Result<skills::SkillRepoSummary, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("skill_repo_upsert", move || {
        skills::repo_upsert(&app, repo_id, &git_url, &branch, enabled)
    })
    .await
}

#[tauri::command]
pub(crate) async fn skill_repo_delete(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    repo_id: i64,
) -> Result<bool, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("skill_repo_delete", move || {
        skills::repo_delete(&app, repo_id)?;
        Ok(true)
    })
    .await
}

#[tauri::command]
pub(crate) async fn skills_installed_list(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
) -> Result<Vec<skills::InstalledSkillSummary>, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("skills_installed_list", move || {
        skills::installed_list(&app)
    })
    .await
}

#[tauri::command]
pub(crate) async fn skills_discover_available(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    refresh: bool,
) -> Result<Vec<skills::AvailableSkillSummary>, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    tauri::async_runtime::spawn_blocking(move || skills::discover_available(&app, refresh))
        .await
        .map_err(|e| format!("SKILL_TASK_JOIN: {e}"))?
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub(crate) async fn skill_install(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    git_url: String,
    branch: String,
    source_subdir: String,
    enabled_claude: bool,
    enabled_codex: bool,
    enabled_gemini: bool,
) -> Result<skills::InstalledSkillSummary, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    tauri::async_runtime::spawn_blocking(move || {
        skills::install(
            &app,
            &git_url,
            &branch,
            &source_subdir,
            enabled_claude,
            enabled_codex,
            enabled_gemini,
        )
    })
    .await
    .map_err(|e| format!("SKILL_TASK_JOIN: {e}"))?
}

#[tauri::command]
pub(crate) async fn skill_set_enabled(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    skill_id: i64,
    cli_key: String,
    enabled: bool,
) -> Result<skills::InstalledSkillSummary, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    tauri::async_runtime::spawn_blocking(move || {
        skills::set_enabled(&app, skill_id, &cli_key, enabled)
    })
    .await
    .map_err(|e| format!("SKILL_TASK_JOIN: {e}"))?
}

#[tauri::command]
pub(crate) async fn skill_uninstall(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    skill_id: i64,
) -> Result<bool, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    tauri::async_runtime::spawn_blocking(move || skills::uninstall(&app, skill_id))
        .await
        .map_err(|e| format!("SKILL_TASK_JOIN: {e}"))??;
    Ok(true)
}

#[tauri::command]
pub(crate) async fn skills_local_list(
    app: tauri::AppHandle,
    cli_key: String,
) -> Result<Vec<skills::LocalSkillSummary>, String> {
    tauri::async_runtime::spawn_blocking(move || skills::local_list(&app, &cli_key))
        .await
        .map_err(|e| format!("SKILL_TASK_JOIN: {e}"))?
}

#[tauri::command]
pub(crate) async fn skill_import_local(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    cli_key: String,
    dir_name: String,
) -> Result<skills::InstalledSkillSummary, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    tauri::async_runtime::spawn_blocking(move || skills::import_local(&app, &cli_key, &dir_name))
        .await
        .map_err(|e| format!("SKILL_TASK_JOIN: {e}"))?
}

#[tauri::command]
pub(crate) async fn skills_paths_get(
    app: tauri::AppHandle,
    cli_key: String,
) -> Result<skills::SkillsPaths, String> {
    blocking::run("skills_paths_get", move || {
        skills::paths_get(&app, &cli_key)
    })
    .await
}
